app/src/main/java/org/greenstand/android/TreeTracker/models/messages/MessagesRepo.kt

```diff
@@ -146,24 +152,46 @@ class MessagesRepo(
             limit = query.limit,
         )
         query = result.query
+        if (query.total == 0) return@withContext
+
+        val saveMessageResponseJobs = mutableListOf<Job>()
+
+        result.messages.forEach {
+            ensureActive()
+            // launch all jobs in parallel and add these to a list.
+            saveMessageResponseJobs.add(launch {
+                saveMessageResponse(wallet, it.copy())
+            })
+        }
 
-        if (query.total == 0) return
+        // wait for all the jobs to complete
+        saveMessageResponseJobs.joinAll()
 
-        result.messages.forEach { saveMessageResponse(wallet, it) }
         while (query.total >= query.limit + query.offset) {
+            ensureActive()
             query = result.query.copy(offset = result.query.offset + result.query.limit)
             result = apiService.getMessages(
                 wallet = wallet,
                 lastSyncTime = lastSyncTime,
                 offset = query.offset,
                 limit = query.limit,
             )
-            result.messages.forEach { saveMessageResponse(wallet, it) }
+
+            saveMessageResponseJobs.clear()
+
+            result.messages.forEach {
```

Comment:
These are small/fast DB transactions. I'm wondering if it's worth adding complexity for a most likely non noticeable performance improvement.